test/controllers/tasks_controller_test.rb

```diff
@@ -88,37 +84,89 @@
   # Unskip and complete these tests for Wave 3
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      get edit_task_path(task.id)
+      must_respond_with :success
     end
     
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      get edit_task_path(-1)
+      must_respond_with :redirect
     end
   end
   
   # Uncomment and complete these tests for Wave 3
   describe "update" do
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
+    before do
+      Task.create(
+        name: "Buy Flowers",
+        description: "Buy flowers that has different colors",
+        completed_at: "today"
+      )
+    end
+    let (:new_task_hash) {
+      {
+        task: {
+          name: "Buy a Book",
+          description: "Buy a book about six sigma",
+          completed_at: "today"
+        },
+      }
+    }
+
     it "can update an existing task" do
-      # Your code here
+      task = Task.first
+      expect {
+        patch task_path(task.id), params: new_task_hash
+      }.wont_change "Task.count"
+
+      must_redirect_to task_path(task.id)
+
+      task = Task.find_by(id: task.id)
+      expect(task.name).must_equal new_task_hash[:task][:name]
+      expect(task.description).must_equal new_task_hash[:task][:description]
+      expect(task.completed_at).must_equal new_task_hash[:task][:completed_at]
     end
     
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      get update_task_path(-1)
+      must_redirect_to root_path
     end
   end
   
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "can delete a task" do
+      task = Task.create(
+        name: "Buy Flowers",
+        description: "Buy flowers that has different colors",
+        completed_at: "today"
+      )
+      expect{
+        delete delete_task_path(task.id)
+      }.must_differ 'Task.count', -1
+    end
     
   end
   
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "can toggle" do
+    task = Task.create(
+      name: "Buy Flowers",
+      description: "Buy flowers that has different colors",
+      completed_at: "today"
+    )
+    
+    patch uncomplete_task_path(task.id)
+    task = Task.find_by(id: task.id)
+    expect(task.completed_at).must_be_nil
+
+    patch complete_task_path(task.id)
+    task = Task.find_by(id: task.id)
+    expect(task.completed_at).wont_be_nil
```

Comment:
I love this test! I think it's pretty awesome that you did multiple "act"s and multiple "assert"s. Usually I'd like to discourage that, but in this case, the asserts are very related, so it reads very well. :)